Add message support

rogeruiz opened the pull request:
I also took some liberties to clean up the code a bit. I'm following the contribution guidelines for Flyjs, but I felt the need to add some more spacing as well. Let me know what you think and I can revert the whitespace edits. Thanks!

rogeruiz committed:
Fix message output by changing `assign` order;

Closes #3

MadcapJake commented:
What do you mean by `message support`?  It already supports messages.  

I thought that pngs were blurry for you.  They're blurry for me, I don't want to change to pngs without completing one of the solutions I mentioned in #4.

Also, I am actually not a fan of the whitespace.  I'm okay with empty lines between the exported function, requires, and the `devicons` function (basically global-only) but everything else should be condensed.  Also, I don't like spaces inside the parens.  I do however use that when creating single-line objects.

MadcapJake commented:
Also, we should stick with not using `var` and actually there really is no need to assign `defaults` as a variable when it's only referenced one time.

All that needs to be done is switch the order of the objects in `assign` right?

rogeruiz commented:
Yup! I just realized that this commit is a mess. An amalgamation of ideas. Sorry about that. 

I'll update this to reflect that and squash it down to the new commit in a few hours. Thanks!

rogeruiz commented:
Alright, I've taken your comments and updated the commit whitespace- and variable-declaration- wise.

I want to write a test for this before merging it in. I'll be able to finish that up later today and I'll ping you either on here or Gitter to let you know when it's done. 🖖

MadcapJake commented:
:+1:  looks great!

rogeruiz committed:
Update eslint;

I'm including `comma-dangle` since it's good practice when working on something
you're going to actually go back to. Also, including `no-use-before-define`
because this was causing an issue with `devicon` because it's _on by default_.

rogeruiz committed:
Add tests

rogeruiz committed:
Add myself to contributors list

rogeruiz committed:
Add `noop` functions for `error` & `debug` methods

rogeruiz commented:
Hey @MadcapJake 

I've added the tests to let me sleep at night about having this PR merged. HAHA.

If @bucaran could weigh in on the best way to test `fly-plugins` then I would feel little about [what I have done here](https://github.com/MadcapJake/fly-notify/commit/da851f381652b863b47f031300f54ffccedec8a7?diff=split).

I added it because Travis-CI was [complaining here](https://travis-ci.org/MadcapJake/fly-notify/jobs/73234711#L178-L189) about `undefined` methods.

MadcapJake commented:
@rogeruiz Fly testing is supposed to land this week so I think we can call this good and modify it as needed when that lands.  Thanks so much for your contribution!

ghost commented:
Promise! :grinning:

rogeruiz commented on test/test.js:
```diff
@@ -1,5 +1,52 @@
+  t.end()
+})
+
+test("Testing bad options", function(t) {
```
Once the test framework lands, I'll refactor this to actually be useful! Haha ¯_(ツ)_/¯

MadcapJake replied:
No fly plugins right now have any meaningful tests so no worries! :+1:

bucaran replied:
Looks basically good to me. @rogeruiz If by framework you mean guidelines, then we are cool. Basically, Fly tests whether filters work, so no fly plugin should run fly and check whether the filter works, instead they should check whether the transform function inside the filter produces the correct output. Most plugins are thin wrappers around any given transform library so these tests should be trivial. I can imagine they will be more complicated for plugins other than transforms/writers.

rogeruiz replied:
@bucaran okay awesome. I like this lean testing idea you want to keep. The gulp tests are insane. When I wrote framework, I meant more convention. Instead of including fly, maybe if we could have something like Fly.util for tests?

bucaran replied:
Yes. Adding a lot of tests does not make your app more safe. Adding the right tests does.

See the tests for `fly-babel`:

```js
const test = require("tape").test
const bind = require("fly-util").bind
const babel = require(bind("../"))

test("fly-babel", (t) => {
  t.plan(2)
  babel.call({
    filter: function (name, transform) {
      t.equal(name, "babel", "add babel filter")
      t.ok(/var a/.test(transform("let a = 0")), "babel transform")
    }
  })
})
```

This is almost enough. Also need to check it is returning the correct extension. But other than that, that's all the plugin should test for. Fly already tests the behavior `Fly.proto.filter`.

bucaran replied:
Notice that

```js
const babel = require(bind("../"))
```

You only need to do that if your plugin is written in ES6/7.

Also notice we are "hijacking" `babel` with our own object that has a mock of `fly` with `filter`.
